### PR TITLE
Dashboards: Fix frontend exception when backend fails to save dashboard

### PR DESCRIPTION
**Pull request by joshhunt**
https://github.com/grafana/grafana/pull/70249 introduced a regression when Grafana is running with the nestedFolders feature flag, and the backend returns an error when saving a dashboard, the frontend fails to handle the error response correctly and it throws a "result is undefined" exception in a toast

image

You can reproduce this by:
 - enable the `nestedFolders` feature flag
 - create a folder at the root named 'My test folder'
 - create a new dashboard and save it at the root as the name `My test folder`
 - observe that the save drawer remains open, and a "result is undefined" (or similar) toast is shown

The expected behaviour (that happens when `nestedFolders` feature flag is not enabled) is that:
 - no "result is undefined" exception is thrown
 - the drawer is closed when the "Dashboard name cannot be the same as folder" toast is shown
 - the drawer doesn't reopen when clicking save again - fixed by https://github.com/grafana/grafana/pull/70434 

This PR fixes this issue :)

### CODE REVIEW

File: public/app/features/dashboard/components/SaveDashboard/useDashboardSave.tsx [hunk 1]
```diff
@@ -14,7 +14,27 @@ import { DashboardSavedEvent } from 'app/types/events';
...
+    const query = await saveDashboardRtkQuery({
+      dashboard: saveModel,
+      folderUid: options.folderUid ?? dashboard.meta.folderUid ?? saveModel.meta.folderUid,
+      message: options.message,
+      overwrite: options.overwrite,
+    });
+
+    if ('error' in query) {
+      throw query.error;
+    }
```
Comment (joshhunt) [inline public/app/features/dashboard/components/SaveDashboard/useDashboardSave.tsx:33]: This is the actual fix - throwing fetch error so it's caught elsewhere.

**Review by axelavargas — COMMENTED** [review]
Great job, @joshhunt 😄 , I gave it a quick test, and with the feature flag enabled it works as expected. But, I found a minor issue when I disabled it:

Whenever I tried to save a dashboard with the same folder name, I get the exception, but then the "Save" button does not open again.

https://github.com/grafana/grafana/assets/239999/875ff4df-d9d4-4ebb-937a-49c079b3b958

**Comment by joshhunt** [comment]
Thanks for checking it out @axelavargas.

It is/was 'expected' that the server error causes the save drawer to no longer open as that was an existing error in main fixed by https://github.com/grafana/grafana/pull/70434. Now that that PR is merged, I've rebased this one.

**Comment by axelavargas** [comment]
> Thanks for checking it out @axelavargas.
> 
> It is/was 'expected' that the server error causes the save drawer to no longer open as that was an existing error in main fixed by #70434. Now that that PR is merged, I've rebased this one.

Awesome, I think I missed this point in the description 😄.

**Review by axelavargas — APPROVED** [review]

**Pull request merged**